### ISSUE TITLE
Performance appears to degrade faster with vo=gpu-next than with vo=gpu + --gpu-api=vulkan

### ISSUE THREAD

**Issue by Dudemanguy** [mpv-player/mpv#9430]
### Important Information

- mpv version: 6952707
- libplacebo version: [eef1105c](https://code.videolan.org/videolan/libplacebo/-/commit/eef1105ce89a388053ed31aca4b92181fbbf6e7e)
- Window Manager and version: Sway 1.6.1 (but it also happens in xorg with i3 4.20.1)
- GPU driver and version: RX 550 using amdgpu with mesa's radv (21.2.5)

### Reproduction steps

Below is my config (essentially only for benchmarking).
```
gpu-api=vulkan
audio=no
untimed=yes
wayland-disable-vsync=yes
video-sync=display-desync
vulkan-swap-mode=mailbox
osd-msg1="FPS: ${estimated-display-fps}"
```
Now, open the same video file but change the vo to compare the performance between the two (`mpv --vo=gpu-next /path/to/file`) and so on. In my case, these are all played at fullscreen (4k).

### Expected behavior

No significant difference.

### Actual behavior

With certain settings, `gpu-next` is considerably slower on my machine. I'll describe the scenarios below with very unscientific measurements (i.e. eyeballing the estimated display fps).

`mpv /path/to/file` (aka no special settings at all besides the config):
**vo=gpu**: Roughly around 745 fps.
**vo=gpu-next**: Roughly around 735 fps.
So in this particular case, the difference looks negligible. However if you throw some settings at it, it degrades fast.

`mpv --profile=gpu-hq /path/to/file`:
**vo=gpu**: Roughly around 226 fps.
**vo=gpu-next**: Roughly around 175 fps.
That's a very big drop. Not really expected. Now this time let's play with deband-iterations to really tax it.

`mpv --deband --deband-iterations=16 /path/to/file`:
**vo=gpu**: Roughly around 60fps.
**vo=gpu-next**: Roughly around 5fps (!).

### Log file

Here are the logs for the `deband-iterations=16` case.
[vo=gpu](https://0x0.st/-5gN.log)
[vo=gpu-next](https://0x0.st/-5gq.log)

**Comment by dixie-flatliner** [comment]
Windows 11, NVIDIA GTX 1660 Ti, driver 496.13, all clocks forced to maximum, 1080p test file:

Defaults:
vo=gpu: ~1200 FPS
gpu-next: ~900 FPS (!)

profile=gpu-hq:
vo=gpu: ~800 FPS
vo=gpu-next: ~650 FPS

deband with 16 passes:
vo=gpu: ~160 FPS
vo=gpu-next: ~110 FPS

At least in my case the numbers are all quite variable, but gpu-next is definitely slower by a decent margin in all of them.

I will say that (in my case), render times are about the same between the two with normal playback settings.

**Comment by haasn** [comment]
Can you test if this is a regression caused by haasn/libplacebo@58cd6f183f?

**Comment by haasn** [comment]
I can actually reproduce your slowdown on my end too (using the bench profile), even though the shader timings are significantly lower for libplacebo on my end.

That would seem to indicate that the difference in performance comes from some thing else, besides the generated shaders. Which leaves astonishingly little wiggle room, given that they use the same underlying vulkan codebase and it's not a result of different shaders.

**Comment by haasn** [comment]
Actually, funnily enough, for me it's opposite: `--vo=gpu` is faster on my default settings, but `--vo=gpu-hq` is way faster with `--deband-iterations=16`.

I'm still curious though what's causing the difference, because they shouldn't be that dramatically different given what they're actually *doing* is still pretty much the same shader code (and I've already ruled out the obvious candidates for the difference).

**Comment by haasn** [comment]
What certainly seems to be the case is that this performance difference, at least for me, only manifests itself when using such a 'bench' configuration.

Comparing my stock settings at normal playback speed, for example, the pass timings are around 4.3ms for a fresh frame using `--vo=gpu-next`, and roughly 9.0ms (of which ~2.7ms are uploading the frame) using `--vo=gpu`. A difference of almost 50% in favor of `gpu-next` (!)

Somehow this difference in timings disappears (and `--vo=gpu-next` picks up an extra slowdown) when using the bench profile? Maybe it's related to a difference in how rendering commands are submitted to the GPU?

**Comment by Dudemanguy** [comment]
I can hit it with normal-esque usage. For example, using FSRCNNX_x2_8-0-4-1.glsl + `gpu-hq` + `display-resample` will drop frames for me with `gpu-next` but not `gpu`.

**Comment by haasn** [comment]
I'll have to look into it further tomorrow. Some observations:

1. I noticed that libplacebo will by default emit code for `--alpha=blend-tiles`, even on files without alpha channels (mpv does not). Removing this (`--alpha=no`) gets rid of a very slight performance discrepancy in the shader.

2. For the out-of-the-box test with default vulkan context settings, the commit I linked earlier is indeed a regression. If I remove it (thus making `gpu-next` prefer fragment shaders) I actually get *better* performance out of `--vo=gpu-next` on stock `--no-config` settings again. So the discrepancy is related to that.

3. Adding just `--deband=yes` to stock settings severely hits performance. I'll re-investigate if it's related to the PRNG, tomorrow.

**Comment by Dudemanguy** [comment]
>I noticed that libplacebo will by default emit code for --alpha=blend-tiles

Hah, that probably explains why sway always has transparent backgrounds. I had not gotten around to reporting/investigating that one yet.

**Comment by haasn** [comment]
> Hah, that probably explains why sway always has transparent backgrounds. I had not gotten around to reporting/investigating that one yet.

That one is probably fixed by haasn/libplacebo@0aaced15da9b9d8950c32be4bff98893b6d8af0f

**Comment by haasn** [comment]
Update re: thread:

I reverted the `compute` upgrading and also fixed the constant-on blend-tiles. So those two performance regressions should be gone. I still haven't looked into the PRNG, but I suspect that's what causes deband to be slower.

Is there any configuration in which you still measure a performance regression with `--deband=no`?

**Comment by Dudemanguy** [comment]
That commit definitely does help. Without any settings, gpu-next is now faster (roughly 927fps vs 753fps). Unfortunately, I still see slowdowns that aren't related to debanding. Selecting a different scaling algorithm will hit it. For example with `mpv --scale=spline36` I get about 283 fps with vo=gpu and about 173 fps with vo=gpu-next.

I still get the transparency thing. I'll open up a separate issue for that.

**Comment by yeezylife** [comment]
`Gpu-next` is generally slower than `gpu` with high consumption profile settings,such as Fsrcnnx,SSimsuperres or high iterations deband.

**Comment by haasn** [comment]
Yeah, confirming that I get a *severe* slowdown (almost factor of 2) just from `--no-config --glsl-shader=FSRCNNX.glsl` (and the appropriate vo).

That means there's definitely _something_ fishy going on, because the shader code is **literally** the same for this shader (and the other shaders are just a no-op, in fact `--vo=gpu` even has one completely unnecessary shader).

I'll look into it further tomorrow, but this is a good reproducing case. It might be related to some constant overhead per pass being executed (since FSRCNN has many passes). So even though the passes are themselves the exact same, something else around them might be different.

**Comment by yeezylife** [comment]
So it's not just about deband(will slow it donw) like the previous speculation.Basiclly anything that make mpv more power consuming.

**Comment by haasn** [comment]
So, I tried looking at `deband` specifically, to see if I could notice anything funny going on. With `--deband-iterations=16` the performance difference is clearly reproducible - almost an order of magnitude. What I noticed:

* With `deband-iterations=1`, gpu-next is slower than gpu by about 50%
* With `deband-iterations=4`, that increases to almost 250%
* With `deband-iterations=16`, the worst case scenario, the difference is a staggering **480%**

So clearly, whatever is causing issues with `gpu-next`, responds in some proportional way to the number of debanding passes being done. This means it must be fundamentally connected to either the PRNG invocation, or texture sampling.

I repeated the test, but with the PRNG commented out (return constant 0.5) from both mpv and libplacebo:

* With `deband-iterations=1`, gpu-next is 20% slower
* With `deband-iterations=4`, that increases to 101%
* With `deband-iterations=16` it increases to 290%

So while the PRNG does appear to be slowing libplacebo down further than mpv (but see below), it's clearly not the only thing to blame. Somehow, the number of texture samples on its own also degrades performance faster on gpu-next (which is the only other result of increasing the deband iterations).

With the libplacebo PRNG modified to use the exact same PRNG as mpv, I get these results:

* `deband-iterations=1`: 45% slowdown
* `deband-iterations=4:` 250% slowdown
* `deband-iterations=16:` 460% slowdown

So the PRNG difference isn't the cause of the issue, but using a PRNG at all definitely exacerbates it. I suspect this is because the PRNG is what makes the dither pattern actually random - note that using a constant function as the PRNG means all passes sample the exact same spot in the texture.

So far, though, it appears as though the core issue here is related to the number of texture fetches the pass is doing, which would explain why it disproportionately hits shaders such as `--deband` and e.g. FSRCNN. Indeed, something like `--scale=ewa_lanczos --scale-radius=4` also reproduces a big slowdown (~90%).

(I also concluded that, testing `gpu-next` against itself, the new PRNG is indeed faster than the old PRNG, and also higher quality, so I'm keeping it as-is)

**Comment by haasn** [comment]
I also noticed that the slowdown is also clearly visible in the pass timers, e.g. the ~3.5ms+2.5ms of luma+chroma debanding in mpv jumps up to ~30ms of debanding in the `gpu-next` implementation. So whatever the difference is here, it's here clearly the result of the **shader itself** executing more slowly on `gpu-next` than `gpu`.

(Which, notably, is unlike the result on FSRCNN, where the pass timers remain the same but things still execute more slowly on `gpu-next` for some reason)

**Comment by haasn** [comment]
Ah, I think the deband thing might be caused by a difference in the way `gpu-next` handles chroma debanding - it effectively tried doing bilinear sampling at the same time, which actually ended up scaling the chroma plane *before* debanding, whereas mpv always does it the other way around. That would explain the performance difference for deband in particular. I'll change the way this is handled.

**Comment by haasn** [comment]
Another thing I noticed: libplacebo by default merges the luma debanding pass in with the chroma scaling pass. This generally cuts down on the number of passes required in total, but apparently, it has the unfortunate side effect of increasing the length of the critical path. I notice this quite clearly in the pass timers: having both in the same pass bumps the luma debanding up from ~4ms to an incredible ~10ms or so, even though the only thing being done in addition to luma debanding is a single `texture()` call on the chroma plane.

When I change it to also insert an indirection in between the luma debanding and the chroma scaling passes, `gpu-next` gets a *lot* faster as a result. However, very surprisingly, running the same configuration through `ffmpeg` has the opposite result, with the merged passes being slightly better.

Somehow, the approach that's faster in `ffmpeg` (and offline processing in general), is significantly slower when used in `mpv`?!

**Comment by haasn** [comment]
> Somehow, the approach that's faster in `ffmpeg` (and offline processing in general), is significantly slower when used in `mpv`?!

Ah, pass merging has nothing to do with it (and is still a good thing to do), it's actually the exact same issue as [this comment](https://github.com/mpv-player/mpv/issues/9430#issuecomment-968110915), when you enable `--deband` without setting any sort of scaler (e.g. `bicubic_fast`), the debanding is now actually done at the *output resolution*, rather than the luma resolution, for the exact same reason. (My forgetting that "free bilinear scaling" only applies to the input, not the output, of a pass)

I've fixed that in haasn/libplacebo@72cd260a, which now once again makes the `gpu-next` pass timings significantly lower than the `gpu` pass timings on my end, even with `--deband=yes --deband-iterations=16`, 4.6ms for gpu-next vs 6.5ms for gpu.

That said, even though the pass timings are lower, the **issue remains** where gpu-next in the bench profile doesn't perform as well as gpu. (Which I'm still not sure is an actual performance issue or just some unrelated difference in the way frames are being fed to these two VOs)

**Comment by haasn** [comment]
Okay, I finally made some progress on this, by running mpv through a GPU profiler. For starters:

1. The lower pass timings in `--vo=gpu-next` correspond to reduced GPU load, reduced GPU power consumption, and reduced GPU activity time. So at the very least, `gpu-next` *is*, in fact, faster, performance-wise. **Meaning the difference we're measuring is coming from somewhere else**
2. Even in `--profile=bench` mode, `--vo=gpu-next` is actually pushing *more* frames per second than `--vo=gpu`, looking at the raw GPU execution of command buffers.
3. Somehow, using `--video-sync=display-desync` is what causes `--vo=gpu-next` to appear to be significantly slower than `--vo=gpu`.
4. Switching to `--video-sync=audio` fixes this, and brings the bench FPS in line with the actual execution times (i.e. slightly faster on `--vo=gpu-next`)

So my conclusions so far are:

1. This issue (after the above issues related to debanding and alpha blending have been fixed) appears to be an illusion of some sort, `gpu-next` *is* actually more performant on a frame-by-frame basis. The difference in performance has something to do with how display sync is handled between the two VOs. It might be that, somehow `vo_gpu` is somehow caching and re-drawing frames where `vo_gpu_next` is not. I'll try and investigate this more specifically further, but at reduced priority.
2. If in doubt, the pass timings are very accurate for vulkan-based VOs. I currently trust them more than the bench profile, for this exact reason.

I'm going to tentatively close this because it seems that the main issue is actually a non-issue, in particular now that the debanding stuff has been fixed. If you notice more, specific performance regressions (as measured by the **pass timers**) for specific configurations, please open separate issues for them.

**Comment by haasn** [comment]
Ah, yes, I figured it out. It was incredibly obvious in hindsight.

If you combine `--video-sync=display-desync` with a 24 fps file on a 60 fps display, mpv will actually draw a 3:2 frame pattern, rather than drawing every frame as fast as possible. So mpv *is* redrawing frames in this configuration. (`--video-sync=audio` does not do this, it draws each frame exactly once)

`gpu_next` currently does not redraw single frames from the cache, which is tracked in https://code.videolan.org/videolan/libplacebo/-/issues/169. This does not apply when using e.g. `--interpolation`, since that goes through the interpolation code which *does* cache frames.

**tl;dr** the remaining difference comes from a known limitation that I still plan to address, and it only affects the configuration of using `--video-sync=display-*` + `--interpolation=no`, and watching a file with less FPS than your display (e.g. 24 fps file on 60 fps display). If you use interpolation, or you don't use display sync, you shouldn't be affected by this regression.

**Comment by Dudemanguy** [comment]
Uh duh, yeah. That would explain it indeed. Thanks for looking into this.
